Attempt to handle `except` cases in a somewhat better way

marcus-oscarsson opened the pull request:
This PR tries to do something with all the `except` cases that may potentially hide an issue.

Replaced all pass statements with a `exception` logger statement and inserted a a `exception` logger statement before return statements.

marcus-oscarsson committed:
Replaced all pass statements with a exception logger statement and nserted a a exception logger statement before return statements

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/sample_changer/Crims.py:
```diff
@@ -16,6 +16,8 @@ def get_image(url: str) -> Optional[bytes]:
             "Failed to fetch image from %s: %s", url, e
         )
+
+        logging.getLogger("HWR").exception("")
```
I write it only here but I guess this applies to all hardware objects, right?

```suggestion
        self.log.exception("")
```

marcus-oscarsson replied:
That will be done in a second pass as we are not sure that all the statements belong to a HardwareObject ...

fabcor-maxiv reviewed (commented):
Great! Much needed! Should hopefully make troubleshooting much easier.

fabcor-maxiv commented:
I think this might make it possible to remove a bunch of linter exceptions (in the "per file" lists).

marcus-oscarsson commented:
This will probably trigger a bunch of other linting errors that I'm not sure how we should handle at the moment ...

marcus-oscarsson commented:
Please have a look, this PR simply changes the except case in the following way:
 - Replaces pass statements with a `exception` logger statement
 - inserts a `exception` logger statement before a return
 - import logging if not already done

We can do a `second` pass to go from `logging.getLogger` to `self.log` in a followup PR

marcus-oscarsson commented:
I would also like to replace the use of the `traceback` module ...

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/sample_changer/Crims.py:
```diff
@@ -147,6 +151,8 @@ def get_processing_plan(
         logging.getLogger("user_level_log").warning(
             "Error getting processing plan from  %s : %s", url, e
         )
+
+        logging.getLogger("HWR").exception("")
```
```suggestion
```

Maybe this one is not necessary. Exception handling seems under control, but of course I might not have all the details.

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/sample_changer/Crims.py:
```diff
@@ -28,6 +30,8 @@ def get_image_size(url: str) -> Tuple[int, int]:
         logging.getLogger("user_level_log").warning(
             "Failed to get image size from %s: %s", url, e
         )
+
+        logging.getLogger("HWR").exception("")
```
```suggestion
```

Maybe this one is not necessary. Exception handling seems under control, but of course I might not have all the details.

fabcor-maxiv commented on mxcubecore/HardwareObjects/abstract/AbstractVideoDevice.py:
```diff
@@ -42,7 +42,7 @@
     import cv2
 except ImportError:
-    pass
+    logging.getLogger("HWR").exception("")
```
```suggestion
```

Maybe this one is not necessary. Exception handling seems under control, but of course I might not have all the details.

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/DigitalZoomMotor.py:
```diff
@@ -55,6 +55,8 @@ def init(self):
             self.camera = self.get_object_by_role("camera")
         except KeyError:
             logging.getLogger("HWR").warning("DigitalZoomMotor: camera not defined")
+
+            logging.getLogger("HWR").exception("")
```
```suggestion
            logging.getLogger("HWR").exception("DigitalZoomMotor: camera not defined")
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11Collect.py:
```diff
@@ -1137,6 +1137,8 @@ def check_path(self, path=None, force=False):
                         os.mkdir(path, mode=0o777)
                     except RuntimeError:
                         self.log.debug("mkdir failed:", str(sys.exc_info()))
+
+                        logging.getLogger("HWR").exception("")
```
```suggestion
                        self.log.exception("mkdir failed: %s", str(sys.exc_info()))
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11MachineInfo.py:
```diff
@@ -98,6 +98,8 @@ def get_value(self):
 
         except Exception as e:
             logging.error(f"Error getting machine values: {e}")
+
+            logging.getLogger("HWR").exception("")
             return {}  # Return empty dictionary if there's an error
```
```suggestion
        except Exception:
            logging.exception("Error getting machine values")
            return {}  # Return empty dictionary if there's an error
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11MachineInfo.py:
```diff
@@ -108,6 +110,8 @@ def get_current(self):
             return round(current_value, 2)
         except AttributeError:
             logging.error("Error reading 'current': Attribute 'current' not found.")
+
+            logging.getLogger("HWR").exception("")
```
```suggestion
            logging.getLogger("HWR").exception("Error reading 'current': Attribute 'current' not found.")
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11MachineInfo.py:
```diff
@@ -116,6 +120,8 @@ def get_lifetime(self):
             return round(lifetime_value, 2)
         except AttributeError:
             logging.error("Error reading 'lifetime': Attribute 'lifetime' not found.")
+
+            logging.getLogger("HWR").exception("")
             return 0
```
```suggestion
except AttributeError:
            logging.getLogger("HWR").exception("Error reading 'lifetime': Attribute 'lifetime' not found.")
            return 0
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11MachineInfo.py:
```diff
@@ -124,6 +130,8 @@ def get_maschine_energy(self):
             return round(energy_value, 2)
         except AttributeError:
             logging.error("Error reading 'energy': Attribute 'energy' not found.")
+
+            logging.getLogger("HWR").exception("")
             return 0
```
```suggestion
except AttributeError:
            logging.getLogger("HWR").exception("Error reading 'energy': Attribute 'energy' not found.")
            return 0
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11MachineInfo.py:
```diff
@@ -132,6 +140,8 @@ def get_message(self):
             return message_value
         except AttributeError:
             logging.error("Error reading 'message': Attribute 'message' not found.")
+
+            logging.getLogger("HWR").exception("")
             return "Message unavailable"
```
```suggestion
        except AttributeError:
            logging.getLogger("HWR").exception("Error reading 'message': Attribute 'message' not found.")
            return "Message unavailable"
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/DESY/P11NanoDiff.py:
```diff
@@ -1473,6 +1473,8 @@ def move_motors(self, motor_positions, timeout=10):
                 self.log.rror(
                     "motor_positions is not a dict and cannot be converted using as_dict()"
                 )
+
+                logging.getLogger("HWR").exception("")
```
```suggestion
                self.log.exception(
                    "motor_positions is not a dict and cannot be converted using as_dict()"
                )
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/EMBL/EMBLBeamlineTest.py:
```diff
@@ -116,6 +116,8 @@ def start_test_queue(self, test_list, create_report=True):
                 logging.getLogger("HWR").warning(
                     "BeamlineTest: Unable to create test directories"
                 )
+
+                logging.getLogger("HWR").exception("")
```
```suggestion
                logging.getLogger("HWR").exception(
                    "BeamlineTest: Unable to create test directories"
                )
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/ESRF/ID29HutchTrigger.py:
```diff
@@ -42,6 +42,8 @@ def init(self):
             self.card, self.channel = map(int, PSSinfo.split("/"))
         except Exception:
             logging.getLogger().error("%s: cannot find PSS number", self.name())
+
+            logging.getLogger("HWR").exception("")
```
```suggestion
            logging.getLogger("HWR").exception("%s: cannot find PSS number", self.name())
```

fabcor-maxiv commented on mxcubecore/HardwareObjects/Gphl/GphlWorkflow.py:
```diff
@@ -1927,6 +1927,8 @@ def calculate_recentring(self, okp, ref_okp, ref_xyz):
                 err.returncode,
                 err.output,
             )
+
+            logging.getLogger("HWR").exception("")
```
```suggestion
            logging.getLogger("HWR").exception(
                "Recen failed with returncode %s. Output was:\n%s",
                err.returncode,
                err.output,
            )
```

fabcor-maxiv reviewed (commented):
OK, I started trying to write fix suggestions, but then realised that this is probably all automatically generated and you might not be interested in manual edits at this stage... I do not know exactly what the plan is. Anyway, while I fully agree with the initiative, I am not 100% convinced that the approach is ideal. In any case, it seems better than doing nothing, but could you tell what the next steps planned are?

rhfogh commented on mxcubecore/HardwareObjects/Gphl/GphlWorkflow.py:
```diff
@@ -1927,6 +1927,8 @@ def calculate_recentring(self, okp, ref_okp, ref_xyz):
                 err.returncode,
                 err.output,
             )
+
```
No, there was already a logger message before the return

marcus-oscarsson replied:
Yes but it does not give us the traceback. I think this exception is actually quite well handled so we can remove the "fix" added by this PR. Its not the case however in most other cases.

rhfogh commented:
I agree with @fabcor-maxiv - this is a very good initiative, but I am not sure it should be merged as it stands. This is clearly done by script, and I notice in at least one case the script has added a log message before a return - even though there was a log message there already. I think someone ought to go through these by hand, one by one, correct as appropriate, and maybe change to get_log (or whatever) It is easier to do it now, when the script output show you exactly where you should look, than it would be to do it after this has been merged.. Simple scripts are very useful, but  - pace Kahnemann ;-)  - they are not a full substitute for human interaction.

marcus-oscarsson commented:
> OK, I started trying to write fix suggestions, but then realised that this is probably all automatically generated and you might not be interested in manual edits at this stage... I do not know exactly what the plan is. Anyway, while I fully agree with the initiative, I am not 100% convinced that the approach is ideal. In any case, it seems better than doing nothing, but could you tell what the next steps planned are?

So yes you are right I did not manually edit all the files, I made a script to do it for me. And yes I partly sold my sole to the devil and asked an AI tool to generate parts of that script. 

The plan is to: 

    1. Replace all the simple cases of pass and return with an exception logger (this PR)
    3. Get rid of `traceback` in `except`
    5. Make use of self.log where applicable 

I still don't know what to do with `error`, `warning`, `info` messages, but at the moment I thought it was better to also ad a separate `exception` log. That's how I read the intent of those logger statements; that the author wanted a separate "user-friendly" log entry. I think that we do on most cases also want a `exception` log.

Id propose that if there are files (or a individual change) that you'd like to remove from the PR that we do so and then we can continue with the next steps.

marcus-oscarsson commented:
> I agree with @fabcor-maxiv - this is a very good initiative, but I am not sure it should be merged as it stands. This is clearly done by script, and I notice in at least one case the script has added a log message before a return - even though there was a log message there already. I think someone ought to go through these by hand, one by one, correct as appropriate, and maybe change to get_log (or whatever) It is easier to do it now, when the script output show you exactly where you should look, than it would be to do it after this has been merged.. Simple scripts are very useful, but - pace Kahnemann ;-) - they are not a full substitute for human interaction.

You are right we need to go through this and remove the cases we don't want "fixed". This goes through all `except` clauses not only the ones for `HardwareObjects` so we can't easily apply the `self.log` construct without making the script quite complex.

What I propose that we do, and that will also take care of the other cases that are still not updated to `self.log` is to make another script that only changes HardwareObjects and make the necessary update.

marcus-oscarsson commented:
I wanted to do these changes in a separate PR but if you prefer I can make the change to `self.log`. It will however be a big change because Ill then change all logger statements.

marcus-oscarsson commented:
Alright I did like this;
 - I made a separate PR for the switch to `self.log` its made to this branch, I think it will make the review easier
 - I removed the cases that guard against import of libraries that does no exist
 - I removed cases where the exception was handled somewhat alright
 - I kept the cases where I thought the exception class is to general for simply logging with `error` or `warning` and I added an additional `exception` log

rhfogh reviewed (approved):
I am happy with this this as it stands now.

And, yes, it is very good that you got this done :-)

elmjag commented:
<img width="735" height="500" alt="image" src="https://github.com/user-attachments/assets/9d886be4-03e0-4867-8802-12377e4d1e24" />

:smile_cat:
